native/components/message-markdown-rules.react.js

```diff
@@ -0,0 +1,136 @@
+// @flow
+
+import type { StyleSheetOf } from '../themes/colors';
+import type { Style } from '../themes/markdown-styles';
+
+import * as React from 'react';
+import { Text } from 'react-native';
+import * as SimpleMarkdown from 'simple-markdown';
+
+export default function rules(
+  styles: StyleSheetOf<Style>,
+  onLink: (url: string) => Promise<any>,
+) {
+  const LINK_INSIDE = '(?:\\[[^\\]]*\\]|[^\\]]|\\](?=[^\\[]*\\]))*';
+  const LINK_HREF_AND_TITLE =
+    '\\s*<?([^\\s]*?)>?(?:\\s+[\'"]([\\s\\S]*?)[\'"])?\\s*';
+  const pressHandler = target => {
+    if (onLink) {
+      onLink(target).catch(error => {
+        console.log(
+          'There has been a problem with this action. ' + error.message,
+        );
+        throw error;
+      });
+    }
+  };
+
+  return {
+    autolink: {
+      ...SimpleMarkdown.defaultRules.autolink,
+      react: (
+        node: SimpleMarkdown.SingleASTNode,
+        output: SimpleMarkdown.Output<SimpleMarkdown.ReactElement>,
+        state: SimpleMarkdown.State,
+      ) => {
+        state.withinText = true;
+        const _pressHandler = () => {
+          pressHandler(node.target);
+        };
+        const autolink = (
+          <Text key={state.key} style={styles.autolink} onPress={_pressHandler}>
+            {output(node.content, state)}
+          </Text>
+        );
+        return autolink;
+      },
+    },
+    link: {
+      ...SimpleMarkdown.defaultRules.link,
+      match: SimpleMarkdown.inlineRegex(
+        new RegExp(
+          '^\\[(' + LINK_INSIDE + ')\\]\\(' + LINK_HREF_AND_TITLE + '\\)',
+        ),
+      ),
+      react: (
+        node: SimpleMarkdown.SingleASTNode,
+        output: SimpleMarkdown.Output<SimpleMarkdown.ReactElement>,
+        state: SimpleMarkdown.State,
+      ) => {
+        state.withinLink = true;
+        const _pressHandler = () => {
+          pressHandler(node.target);
+        };
+        const link = (
+          <Text key={state.key} style={styles.autolink} onPress={_pressHandler}>
+            {output(node.content, state)}
+          </Text>
+        );
+        state.withinLink = false;
+        return link;
+      },
+    },
+    // in some cases, simple-markdown recognizes "normal" text as a paragraph
+    // and we need this rule to preserve our style and bubble size
+    paragraph: {
```

Comment:
I'd like to understand how SimpleMarkdown's parser works better. I still don't understand the difference between "paragraph" and "text", or between "autolink"/"link"/"url".

For each of these, can you give me an example of Markdown syntax that will trigger them? I'd ideally like to also see the corresponding AST that SimpleMarkdown's parser generates from it.

Reply:
Paragraph is like a main node, holding other inside it's content. If content is contains something more than just plain text (for example text, then some link ) then it will be splited into separate nodes, with appropriate types (in this case, text and link). Paragraph is responsible for wrapping message bubble over text.
Text is just a normal text, styled like link or plain text.
Autolink matches `<https://google.com>` syntax, url  `https://www.google.com` (or ` https://www.google.com` or other starting with http), and link will be triggered by `[google link](https://www.google.com)`. It's not needed for us, at least now, but we need to have link since both autolink and url are going through link in `simple-markdown`.
![Zrzut ekranu 2020-07-7 o 11 52 24](https://user-images.githubusercontent.com/18313471/86764942-cc8a1080-c048-11ea-805a-44f72caf7bf0.png)
![Zrzut ekranu 2020-07-7 o 11 52 59](https://user-images.githubusercontent.com/18313471/86764946-cdbb3d80-c048-11ea-8003-81f67c2e21cf.png)
![Zrzut ekranu 2020-07-7 o 11 55 38](https://user-images.githubusercontent.com/18313471/86764949-ce53d400-c048-11ea-9064-47e324cac98d.png)